File: packages/azure/android/src/main/java/com/clarionhq/azure/AzureRecognizerEngine.kt
package com.clarionhq.azure

import com.microsoft.cognitiveservices.speech.CancellationErrorCode
import com.microsoft.cognitiveservices.speech.CancellationReason
import com.microsoft.cognitiveservices.speech.Recognizer
import com.microsoft.cognitiveservices.speech.RecognitionResult
import com.microsoft.cognitiveservices.speech.SpeechRecognizer
import com.microsoft.cognitiveservices.speech.SpeechConfig
import com.microsoft.cognitiveservices.speech.audio.AudioConfig
import com.microsoft.cognitiveservices.speech.transcription.ConversationTranscriber

/**
 * Receives normalized events from either [SpeechRecognizer] or
 * [ConversationTranscriber] so [AzureSession] can stay engine-agnostic.
 */
internal interface AzureRecognizerCallbacks {
  fun azureRecognizing(result: RecognitionResult, speakerId: String)
  fun azureRecognized(result: RecognitionResult, speakerId: String)
  fun azureCanceled(
    reason: CancellationReason?,
    errorCode: CancellationErrorCode?,
    errorDetails: String?,
  )
  fun azureSessionStopped()
  /** `kind` = "started" or "ended", `offsetMs` in ms since session start (-1 if unknown). */
  fun azureSpeechBoundary(kind: String, offsetMs: Double)
}

/**
 * Internal-only abstraction over the two Azure recognizer flavors.
 * Both wrappers expose the same lifecycle so [AzureSession] doesn't have to
 * branch on diarization in every method.
 */
internal interface AzureRecognizerEngine {
  /** Underlying SDK recognizer — exposed for property-bag access and tests. */
  val sdkRecognizer: Recognizer

  fun startAsync(): java.util.concurrent.Future<Void>
  fun stopAsync(): java.util.concurrent.Future<Void>
  fun close()

  /**
   * Hot-swap the auth token on the property bag of the concrete recognizer.
   * Base [Recognizer] doesn't expose this — each impl pokes its own subclass's
   * `properties` collection instead, which the SDK reads on next reconnect.
   */
  fun updateAuthToken(token: String)
}

// region — Speech recognition wrapper (no diarization)

internal class AzureSpeechRecognizerEngineImpl(
  speechConfig: SpeechConfig,
  audioConfig: AudioConfig,
  callbacks: AzureRecognizerCallbacks,
) : AzureRecognizerEngine {

  private val recognizer = SpeechRecognizer(speechConfig, audioConfig)

  override val sdkRecognizer: Recognizer get() = recognizer

  init {
    recognizer.recognizing.addEventListener { _, args ->
      callbacks.azureRecognizing(args.result, speakerId = "")
    }
    recognizer.recognized.addEventListener { _, args ->
      callbacks.azureRecognized(args.result, speakerId = "")
    }
    recognizer.canceled.addEventListener { _, args ->
      callbacks.azureCanceled(args.reason, args.errorCode, args.errorDetails)
    }
    recognizer.sessionStopped.addEventListener { _, _ ->
      callbacks.azureSessionStopped()
    }
    recognizer.speechStartDetected.addEventListener { _, args ->
      val ms = args.offset.toDouble() / 10_000.0
      callbacks.azureSpeechBoundary("started", if (ms >= 0) ms else -1.0)
    }
    recognizer.speechEndDetected.addEventListener { _, args ->
      val ms = args.offset.toDouble() / 10_000.0
      callbacks.azureSpeechBoundary("ended", if (ms >= 0) ms else -1.0)
    }
  }

  override fun startAsync() = recognizer.startContinuousRecognitionAsync()
  override fun stopAsync() = recognizer.stopContinuousRecognitionAsync()
  override fun close() = recognizer.close()
  override fun updateAuthToken(token: String) = pokeAuthToken(recognizer, token)
}

// endregion

// region — Conversation transcriber wrapper (diarization)

internal class AzureConversationTranscriberEngineImpl(
  speechConfig: SpeechConfig,
  audioConfig: AudioConfig,
  callbacks: AzureRecognizerCallbacks,
) : AzureRecognizerEngine {

  private val transcriber = ConversationTranscriber(speechConfig, audioConfig)

  override val sdkRecognizer: Recognizer get() = transcriber

  init {
    transcriber.transcribing.addEventListener { _, args ->
      callbacks.azureRecognizing(args.result, speakerId = args.result.speakerId ?: "")
    }
    transcriber.transcribed.addEventListener { _, args ->
      callbacks.azureRecognized(args.result, speakerId = args.result.speakerId ?: "")
    }
    transcriber.canceled.addEventListener { _, args ->
      callbacks.azureCanceled(args.reason, args.errorCode, args.errorDetails)
    }
    transcriber.sessionStopped.addEventListener { _, _ ->
      callbacks.azureSessionStopped()
    }
    transcriber.speechStartDetected.addEventListener { _, args ->
      val ms = args.offset.toDouble() / 10_000.0
      callbacks.azureSpeechBoundary("started", if (ms >= 0) ms else -1.0)
    }
    transcriber.speechEndDetected.addEventListener { _, args ->
      val ms = args.offset.toDouble() / 10_000.0
      callbacks.azureSpeechBoundary("ended", if (ms >= 0) ms else -1.0)
    }
  }

  override fun startAsync() = transcriber.startTranscribingAsync()
  override fun stopAsync() = transcriber.stopTranscribingAsync()
  override fun close() = transcriber.close()
  override fun updateAuthToken(token: String) = pokeAuthToken(transcriber, token)
}

/**
 * Writes a fresh auth token into a recognizer's property bag. The SDK reads
 * the value on the next websocket reconnect, so this is a no-op for the
 * in-flight session — by design. Base [Recognizer] doesn't expose
 * `properties`; we narrow to the two concrete types we use.
 */
private fun pokeAuthToken(recognizer: Recognizer, token: String) {
  val props = when (recognizer) {
    is com.microsoft.cognitiveservices.speech.SpeechRecognizer -> recognizer.properties
    is com.microsoft.cognitiveservices.speech.transcription.ConversationTranscriber -> recognizer.properties
    else -> return
  }
  props.setProperty(
    com.microsoft.cognitiveservices.speech.PropertyId.SpeechServiceAuthorization_Token,
    token,
  )
}

// endregion

File: packages/azure/ios/AzureRecognizerEngine.swift
import Foundation
import MicrosoftCognitiveServicesSpeech

/// Receives normalized events from either `SPXSpeechRecognizer` or
/// `SPXConversationTranscriber` so `AzureSession` can stay engine-agnostic.
internal protocol AzureRecognizerCallbacks: AnyObject {
  func azureRecognizing(result: SPXRecognitionResult, speakerId: String)
  func azureRecognized(result: SPXRecognitionResult, speakerId: String)
  func azureCanceled(
    reason: SPXCancellationReason,
    errorCode: SPXCancellationErrorCode,
    errorDetails: String?
  )
  func azureSessionStopped()
  /// `kind` = "started" or "ended", `offsetMs` in ms since session start (-1 if unknown).
  func azureSpeechBoundary(kind: String, offsetMs: Double)
}

/// Internal-only abstraction over the two Azure recognizer flavors.
/// Both wrappers expose the same lifecycle so `AzureSession` doesn't have to
/// branch on diarization in every method.
internal protocol AzureRecognizerEngine: AnyObject {
  /// The underlying SPX recognizer instance — exposed so callers can reach
  /// its `properties` bag (e.g. to hot-swap the auth token between sessions).
  var spxBase: SPXRecognizer { get }
  func start() throws
  func stop() throws
}

// MARK: - Speech-recognition wrapper (no diarization)

internal final class AzureSpeechRecognizerEngine: AzureRecognizerEngine {
  private let recognizer: SPXSpeechRecognizer
  private weak var callbacks: AzureRecognizerCallbacks?

  var spxBase: SPXRecognizer { recognizer }

  init(speechConfig: SPXSpeechConfiguration, callbacks: AzureRecognizerCallbacks) throws {
    self.recognizer = try SPXSpeechRecognizer(speechConfig)
    self.callbacks = callbacks
    bindHandlers()
  }

  func start() throws { try recognizer.startContinuousRecognition() }
  func stop() throws { try recognizer.stopContinuousRecognition() }

  private func bindHandlers() {
    recognizer.addRecognizingEventHandler { [weak self] _, args in
      self?.callbacks?.azureRecognizing(result: args.result, speakerId: "")
    }
    recognizer.addRecognizedEventHandler { [weak self] _, args in
      self?.callbacks?.azureRecognized(result: args.result, speakerId: "")
    }
    recognizer.addCanceledEventHandler { [weak self] _, args in
      self?.callbacks?.azureCanceled(
        reason: args.reason,
        errorCode: args.errorCode,
        errorDetails: args.errorDetails
      )
    }
    recognizer.addSessionStoppedEventHandler { [weak self] _, _ in
      self?.callbacks?.azureSessionStopped()
    }
    recognizer.addSpeechStartDetectedEventHandler { [weak self] _, args in
      let offsetMs = Double(args.offset) / 10_000.0
      self?.callbacks?.azureSpeechBoundary(kind: "started", offsetMs: offsetMs >= 0 ? offsetMs : -1)
    }
    recognizer.addSpeechEndDetectedEventHandler { [weak self] _, args in
      let offsetMs = Double(args.offset) / 10_000.0
      self?.callbacks?.azureSpeechBoundary(kind: "ended", offsetMs: offsetMs >= 0 ? offsetMs : -1)
    }
  }
}

// MARK: - Conversation-transcriber wrapper (diarization)

internal final class AzureConversationTranscriberEngine: AzureRecognizerEngine {
  private let transcriber: SPXConversationTranscriber
  private weak var callbacks: AzureRecognizerCallbacks?

  var spxBase: SPXRecognizer { transcriber }

  init(speechConfig: SPXSpeechConfiguration, callbacks: AzureRecognizerCallbacks) throws {
    // SPXConversationTranscriber requires an explicit audio configuration —
    // it doesn't have a default-mic-only init.
    let audioConfig = SPXAudioConfiguration()
    self.transcriber = try SPXConversationTranscriber(
      speechConfiguration: speechConfig,
      audioConfiguration: audioConfig
    )
    self.callbacks = callbacks
    bindHandlers()
  }

  // The Async-suffixed methods take a `(Bool, Error?) -> Void` completion
  // block — we don't need to know when start/stop finishes since events flow
  // on the listener anyway. Just log unexpected errors.
  func start() throws {
    try transcriber.startTranscribingAsync { ok, err in
      if !ok, let err = err { NSLog("[ClarionAzure] startTranscribingAsync failed: \(err)") }
    }
  }
  func stop() throws {
    try transcriber.stopTranscribingAsync { ok, err in
      if !ok, let err = err { NSLog("[ClarionAzure] stopTranscribingAsync failed: \(err)") }
    }
  }

  private func bindHandlers() {
    transcriber.addTranscribingEventHandler { [weak self] _, args in
      guard let result = args.result else { return }
      self?.callbacks?.azureRecognizing(
        result: result,
        speakerId: result.speakerId ?? ""
      )
    }
    transcriber.addTranscribedEventHandler { [weak self] _, args in
      guard let result = args.result else { return }
      self?.callbacks?.azureRecognized(
        result: result,
        speakerId: result.speakerId ?? ""
      )
    }
    transcriber.addCanceledEventHandler { [weak self] _, args in
      self?.callbacks?.azureCanceled(
        reason: args.reason,
        errorCode: args.errorCode,
        errorDetails: args.errorDetails
      )
    }
    transcriber.addSessionStoppedEventHandler { [weak self] _, _ in
      self?.callbacks?.azureSessionStopped()
    }
    transcriber.addSpeechStartDetectedEventHandler { [weak self] _, args in
      let offsetMs = Double(args.offset) / 10_000.0
      self?.callbacks?.azureSpeechBoundary(kind: "started", offsetMs: offsetMs >= 0 ? offsetMs : -1)
    }
    transcriber.addSpeechEndDetectedEventHandler { [weak self] _, args in
      let offsetMs = Double(args.offset) / 10_000.0
      self?.callbacks?.azureSpeechBoundary(kind: "ended", offsetMs: offsetMs >= 0 ? offsetMs : -1)
    }
  }
}
